### PR TITLE
Debugging for big linestring, limit geom size

### PR DESCRIPTION
**Pull request by loicgasser**
Evaluation for https://github.com/geoadmin/mf-chsdi3/issues/2702

`./scripts/test-payload.sh`

5300 coordinates is still fine but it takes up to 10 sec.
It seems that blocking at 150'000 content length would already be quite permissive. do you agree?
ping @gjn @davidoesch 

`Content-Length: 127487`

Then I have to find out why the `@` notation with curl doesn't work. Body is too big to be passed as a string to curl beyond that.

ping @procrastinatio could you give me a hint?
I would like to block it at Apache level.

https://github.com/geoadmin/mf-chsdi3/issues/2702

### CODE REVIEW

**Comment by loicgasser** [comment]
@gjn @procrastinatio I added a check on geom size,  either in body or in parameter we don't accept linestrings bigger than 160'000 characters or approximately 6500 thousands coordinates.

**Comment by loicgasser** [comment]
I optimized the number of threads and processes for the prefork mode.
I believe that if we change the mode, we might as well migrate this service all together. I am sure we can handle much more requests than we do a the moment.

As always I took the best of 5 and tested on an dev (large VS xlarge on prod)
with 300 concurrent requests per second on the height service we get:

```
ab -r -n 3000  -c 300 -H "Host: service-alti.dev.bgdi.ch" "http://127.0.0.1/ltgal/rest/services/height?easting=2721100&elevation_model=COMB&northing=1240250"

-----------------------

Nb Processes: 1
Nb Threads: 15
Failed requests:        0
Requests per second:    157.47 [#/sec] (mean)

Nb Processes: 1
Nb Threads: 30
Failed requests:        0
Requests per second:    153.33 [#/sec] (mean)

Nb Processes: 1
Nb Threads: 45
Failed requests:        0
Requests per second:    153.01 [#/sec] (mean)

-----------------------

Nb Processes: 2
Nb Threads: 15
Failed requests:        0
Requests per second:    257.10 [#/sec] (mean)

Nb Processes: 2
Nb Threads: 30
Failed requests:        0
Requests per second:    252.52 [#/sec] (mean)

Nb Processes: 2
Nb Threads: 45
Failed requests:        0
Requests per second:    261.92 [#/sec] (mean)

-----------------------

Nb Processes: 4
Nb Threads: 15
Failed requests:        0
Requests per second:    280.52 [#/sec] (mean)

Nb Processes: 4
Nb Threads: 30
Failed requests:        0
Requests per second:    266.89 [#/sec] (mean)

Nb Processes: 4
Nb Threads: 45
Failed requests:        0
Requests per second:    259.40 [#/sec] (mean)

-----------------------

Nb Processes: 6
Nb Threads: 15
Failed requests:        0
Requests per second:    281.82 [#/sec] (mean)

Nb Processes: 6
Nb Threads: 30
Failed requests:        0
Requests per second:    268.08 [#/sec] (mean)

Nb Processes: 6
Nb Threads: 45
Failed requests:        0
Requests per second:    254.99 [#/sec] (mean)

--------------------------

Nb Processes: 8
Nb Threads: 15
Failed requests:        0
Requests per second:    274.89 [#/sec] (mean)

Nb Processes: 8
Nb Threads: 30
Failed requests:        0
Requests per second:    278.10 [#/sec] (mean

Nb Processes: 8
Nb Threads: 45
Failed requests:        0
Requests per second:    266.16 [#/sec] (mean)

--------------------------

Nb Processes: 10
Nb Threads: 15
Failed requests:        0
Requests per second:    278.75 [#/sec] (mean)

Nb Processes: 10
Nb Threads: 30
Failed requests:        0
Requests per second:    270.38 [#/sec] (mean)

Nb Processes: 10
Nb Threads: 45
Failed requests:        0
Requests per second:    235.84 [#/sec] (mean)

--------------------------

Nb Processes: 12
Nb Threads: 15
Failed requests:        0
Requests per second:    251.87 [#/sec] (mean)


Nb Processes: 12
Nb Threads: 30
Failed requests:        0
Requests per second:    235.06 [#/sec] (mean)


Nb Processes: 12
Nb Threads: 45
Failed requests:        0
Requests per second:    261.08 [#/sec] (mean)

--------------------------
```

The best results for us being


```
Nb Processes: 4
Nb Threads: 15
Failed requests:        0
Requests per second:    280.52 [#/sec] (mean)


Nb Processes: 6
Nb Threads: 15
Failed requests:        0
Requests per second:    281.82 [#/sec] (mean)

```

So re-tested the 2 best configurations against what we have at the moment (30 threads and 1 process)
and try to see when the service starts to break and we start seeing errors
(I warmed up each service beforehand)


at 390 requests/sec for 30 threads and 1 process (91.53 [#/sec] (mean))
```
ab -r -n 3900  -c 390 -H "Host: service-alti.dev.bgdi.ch" "http://127.0.0.1/ltgal/rest/services/height?easting=2721100&elevation_model=COMB&northing=1240250"
````

at 460  requests/sec for 15 threads and 4 processes (112.09 [#/sec])
```
ab -r -n 4600  -c 460 -H "Host: service-alti.dev.bgdi.ch" "http://127.0.0.1/ltgal/rest/services/height?easting=2721100&elevation_model=COMB&northing=1240250"
```

at 450  requests/sec for 15 threads and 6 processes (131.06 [#/sec] (mean))
```
ab -r -n 4500  -c 450 -H "Host: service-alti.dev.bgdi.ch" "http://127.0.0.1/ltgal/rest/services/height?easting=2721100&elevation_model=COMB&northing=1240250"
```

Since the profile is now alone on its instance, I propose to use:

15 threads and 4 processes to maximize concurrency and still be on the safe side.

**Comment by loicgasser** [comment]
@davidoesch Do you agree to impose a maxmimum number of coordinates one can send to the profile?

I propose `6'000-6'500` max coordinates in this PR. Nowing that this is the upper limit before the request either takes way too long `> 10 sec` or the server crashed. I would of course document that in the new API page.

**Comment by davidoesch** [comment]
Question
- [ ] how many of the generated  KML last 6 month are > 6000 vertexes (and therefore  affected?)
- [ ] Document limit in API heigh/profile
- [ ] Document limit in https://help.geo.admin.ch/?id=94&lang=de ( we then show a message as prposed by oterral to inform user: profile to long)

**Comment by davidoesch** [comment]
Then let me rephrase my question for the front end
- [ ] how many of the generated KML last 6 month are > 5000 vertexes (and therefore affected?)
- [ ]  @davidoesch will Document limit in https://help.geo.admin.ch/?id=94&lang=de ( we then show a message as prposed by oterral to inform user: profile to long)

For the backend : I Agree on the 6000 vertexes if the front end solution is not affecting to many users...

sorry to be picky but I need to make sure if we break some active uses cases of users (hint: freehand draw)

**Comment by oterral** [comment]
@davidoesch I don't see anything on kibana. The request Content-length header is not in the log and we don't see POST parameters.

**Comment by davidoesch** [comment]
then go
lets take the risk (we can roll back if there is a show stopper)
+1

**Comment by loicgasser** [comment]
ok, but I think the risk is minimal beyond 6'000 requests already start failing.

File: alti/subscribers.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+from pyramid.events import subscriber, NewRequest
+from pyramid.httpexceptions import HTTPRequestEntityTooLarge
+
+
+@subscriber(NewRequest)
+def validate_body_length(event):
+    max_clength = 160000
+    content_length = event.request.headers.get('Content-Length')
+    content_length = content_length or len(event.request.params.get('geom', ''))
```
Comment (gjn) [inline alti/subscribers.py]: Does this cover POST requests?

Does this check content-lenght before the requests is parsed (POST body)?

I think we could say that for profile requests, we ignore requests that don't have a content-length header set at all when POST

File: alti/tests/integration/test_profile.py [hunk 2]
```diff
@@ -137,3 +157,8 @@ def test_profile_lv03_json_invalid_offset(self):
         params = {'geom': '{"type":"LineString","coordinates":[[550050,206550],[556950,204150],[561050,207950]]}', 'offset': 'asdf'}
         resp = self.testapp.get('/rest/services/profile.json', params=params, headers=self.headers, status=400)
         resp.mustcontain("Please provide a numerical value for the parameter 'offset'")
+
+    def test_profile_entity_too_large(self):
+        params = {'geom': create_json(7000), 'sr': '2056'}
+        resp = self.testapp.get('/rest/services/profile.json', params=params, headers=self.headers, status=413)
+        resp.mustcontain('LineString too large')
```
Comment (gjn) [inline alti/tests/integration/test_profile.py]: I think we also need to test the same for POST requests.

File: scripts/test-payload.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+source .venv/bin/activate
+python scripts/generate-linestring.py 5000
+PAYLOAD=$(cat payload.json)
+curl -F "geom=${PAYLOAD}" \
+     -F "elevation_models=COMB" \
+     -H "Referer: https://map.geo.admin.ch" \
```
Comment (gjn) [inline scripts/test-payload.sh]: For tests, we should use dedicated referer. That way, we can filter them easily in the logs. something like unittest.geo.admin.ch or test.geo.admin.ch that let's you know the origin of the requests in the logs.

File: alti/subscribers.py (hunk 1 above)
Comment (loicgasser) [reply alti/subscribers.py]: We can use both GET and POST, here we validate both geom and/or length of geom parameter.
> I think we could say that for profile requests, we ignore requests that don't have a content-length header set at all when POST

I can add this check, clients usually implement it.

File: alti/tests/integration/test_profile.py (hunk 2 above)
Comment (loicgasser) [reply alti/tests/integration/test_profile.py]: Yes I add a test

File: scripts/test-payload.sh (hunk 3 above)
Comment (loicgasser) [reply scripts/test-payload.sh]: This was to test locally. Not on prod.

:)

But in general yes makes sense

File: alti/subscribers.py (hunk 1 above)
Comment (gjn) [reply alti/subscribers.py]: I think there are limits on the GET params length anyway, either on apache, varnish or other proxy levels. Heck, even IE has a limit for total URL lenght.

My worry is that we need to parse the content to determine its length - and that the parsing of the content is slow. But I have not tested this.

Comment (loicgasser) [reply alti/subscribers.py]: IE9 has a maxium of 2048.
It really varies according to the browser here is a complete list.
https://boutell.com/newfaq/misc/urllength.html

File: alti/tests/integration/test_profile.py (hunk 2 above)
Comment (loicgasser) [reply alti/tests/integration/test_profile.py]: test added

File: alti/subscribers.py (hunk 1 above)
Comment (loicgasser) [reply alti/subscribers.py]: Note that this happend before entering the view, so no json parsing here, it's before.
I can't do it at Apache level, which was my first attempt.

**Comment by loicgasser** [comment]
now rejecting POST requests without content-length

**Comment by loicgasser** [comment]
https://github.com/geoadmin/mf-chsdi3/compare/gal_doc_prof?expand=1
